improver/synthetic_data/generate_metadata.py

```diff
@@ -0,0 +1,212 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to generate a metadata cube."""
+
+from datetime import datetime, timedelta
+
+import numpy as np
+from iris.std_names import STD_NAMES
+from iris.util import squeeze
+
+from improver.synthetic_data.set_up_test_cubes import (
+    set_up_percentile_cube,
+    set_up_probability_cube,
+    set_up_variable_cube,
+)
+
+DEFAULT_GRID_SPACING = {"latlon": 0.02, "equalarea": 2000}
+CUBE_TYPES = ("variable", "percentile", "probability")
+
+
+def _get_units(name):
+    """ Get output variable units from iris.std_names.STD_NAMES """
+    try:
+        units = STD_NAMES[name]["canonical_units"]
+    except KeyError:
+        raise ValueError("Units of {} are not known.".format(name))
+
+    return units
+
+
+def _create_time_bounds(time, time_period):
+    """ Create time bounds using time - time_period as the lower bound and time as the
+    upper bound"""
+    lower_bound = time - timedelta(minutes=time_period)
+    upper_bound = time
+
+    return (lower_bound, upper_bound)
+
+
+def _create_data_array(ensemble_members, leading_dimension, npoints, height_levels):
+    """ Create data array of specified shape filled with zeros """
+    if leading_dimension is not None:
+        nleading_dimension = len(leading_dimension)
+    else:
+        nleading_dimension = ensemble_members
+
+    if height_levels is not None:
+        nheight_levels = len(height_levels)
+    else:
+        nheight_levels = None
+
+    data_shape = []
+
+    if nleading_dimension > 1:
+        data_shape.append(nleading_dimension)
+    if nheight_levels is not None:
+        data_shape.append(nheight_levels)
+
+    data_shape.append(npoints)
+    data_shape.append(npoints)
+
+    return np.zeros(data_shape, dtype=np.float32)
+
+
+def generate_metadata(
+    name="air_pressure_at_sea_level",
+    units=None,
+    spatial_grid="latlon",
+    time=datetime(2017, 11, 10, 4, 0),
+    time_period=None,
+    ensemble_members=8,
+    leading_dimension=None,
+    cube_type="variable",
+    spp__relative_to_threshold="above",
+    npoints=71,
+    height_levels=None,
+    **kwargs,
+):
```

Comment:
Missing `frt` option.

Reply:
I didn't put them in kwargs because they're required for additional processing in the function. So `time` (or its default value) is required to create `time_bounds` if `time_period` is provided, `spatial_grid` (or its default value) is required to set the default `grid_spacing` if `grid_spacing` is not provided, and `height_levels` (or default of None) is required to create the data array.

Could always put them into kwargs and add the defaults as constants?

```py
    if time_period is not None:
        if "time" not in kwargs:
            kwargs["time"] = DEFAULT_TIME
        time_bounds = _create_time_bounds(kwargs["time"], time_period)
        kwargs["time_bounds"] = time_bounds
```

Reply:
Yes, I prefer that second option.

Reply:
Updated in commit 5e90f39a3ef8f76c7a74a6e3126d5a36b609f17e.

Additional changes in commits 76abdd53c9b819ebe85740b52b2417996033d16c and 043d2ec33db5f6704e5f4a8d18dbad953cbafdc1.